src/tools/chdman.cpp

```diff
@@ -1016,6 +1016,72 @@ static void guess_chs(std::string *filename, uint64_t filesize, int sectorsize,
 }
 
 
+//-------------------------------------------------
+//  reopen_chd_with_parent - find a parent chd in the
+//  directory specified
+//-------------------------------------------------
+
+static chd_error reopen_chd_with_parents(chd_file &input, const char* chd_path, const std::string parents_dirpath, bool writeable = false)
+{
+	if (!input.opened())
+  {
+		input.open(chd_path, writeable);
+	}
+	chd_file *parent = new chd_file;
+	chd_file *child = NULL;
+
+	int depth = 0;
+	std::string paths[8];
+	std::string filepath;
+	util::sha1_t parent_sha1 = input.parent_sha1();
+	chd_error err;
+	while (true)
+	{
+		osd::directory::ptr dir = osd::directory::open(parents_dirpath);
+		const osd::directory::entry *entry;
+		while((entry = dir->read()) != nullptr)
+		{
+			filepath = parents_dirpath + PATH_SEPARATOR + entry->name;
+			err = parent->open(filepath.c_str());
+			if (err != CHDERR_NONE)
+				continue;
+			if (parent->sha1() == parent_sha1)
+      {
+				paths[depth] = filepath;
+				break;
+			}
+			parent->close();
+		}
+		// No parent found in parents_dirpath
+		if (!parent->opened())
+			return CHDERR_REQUIRES_PARENT;
+		if (parent->parent_sha1() == util::sha1_t::null)
+			break;
+
+		depth++;
+		// Don't dive deeper than 8 parents
+		if (depth >= 8)
+			return CHDERR_REQUIRES_PARENT;
+		parent_sha1 = parent->parent_sha1();
+		parent->close();
+	}
+	for (int d = depth-1; d >= 0; d--)
+  {
+		child = new chd_file;
+		err = child->open(paths[d].c_str(), writeable, parent);
```

Comment:
Will this delete the parent even if it fails?  Actually, looking at the code I don’t think it ever deletes the parent.  This is a resource leak.

Reply:
The parent MUST survive after the function returns, so that's why it's never deleted. This is needed, so the child chd_file's m_parent pointer is not invalidated when a read occurs later on.